### PR TITLE
Implement model for String.equals

### PR DESCRIPTION
**Pull request by codebyzeb**
String.equals currently handled by CBMC, this removes support
within CBMC so that String.equals will be handled by a model instead.
Includes a pointer update to the definition change in the
java-models-library submodule.

### CODE REVIEW

**Comment by codebyzeb** [comment]
Pointer update to: diffblue/java-models-library#7

**Comment by tautschnig** [comment]
I am a nominated reviewer, but I'd hope for @romainbrenguier and @cesaro to weigh in/approve.

File: jbmc/regression/strings-smoke-tests/java_value_of_long/test.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 public class test
-{
+
```
Comment (romainbrenguier) [inline jbmc/regression/strings-smoke-tests/java_value_of_long/test.java]: This is probably an involuntary change

File: src/solvers/flattening/boolbv_array.cpp [hunk 2]
```diff
@@ -11,17 +11,16 @@ Author: Daniel Kroening, kroening@kroening.com
 
 bvt boolbvt::convert_array(const exprt &expr)
 {
-  std::size_t width=boolbv_width(expr.type());
+  const std::size_t width = boolbv_width(expr.type());
+  const exprt::operandst &operands = expr.operands();
 
-  if(width==0)
-    return conversion_failed(expr);
+  if(operands.empty() && width == 0)
```
Comment (martin-cs) [inline src/solvers/flattening/boolbv_array.cpp:18]: Is this adding functionality or is this just suppressing an error case?  What is the use-case for this?

**Review by martin-cs — COMMENTED** [review]
I've was asked to review this but there is only one file within my remit so I'm reluctant to approve.

File: src/solvers/flattening/boolbv_array.cpp (hunk 2 above)
Comment (romainbrenguier) [reply src/solvers/flattening/boolbv_array.cpp:18]: This is removing a warning in case there is an equality between empty arrays (this was occurring in some of the tests, when I added the java models library to the classpath)

Comment (martin-cs) [reply src/solvers/flattening/boolbv_array.cpp:18]: OK but why is the library model comparing empty arrays?  Maybe there is a good reason but equally, it could be a bug, right?

Comment (romainbrenguier) [reply src/solvers/flattening/boolbv_array.cpp:18]: I think it's because there are some empty string literals there. I not sure about the exact reason, but I don't see anything wrong with comparing empty arrays, no?

Comment (tautschnig) [reply src/solvers/flattening/boolbv_array.cpp:18]: Is an empty string actually an empty array? I'm with Martin, this seems worth investigating. Maybe there are good reasons, but those reasons haven't come up in the last 15 years and thus should be free from doubt and guesswork.

Comment (martin-cs) [reply src/solvers/flattening/boolbv_array.cpp:18]: @romainbrenguier : I doubt it is intrinsically harmful but I suspect that quite a bit of the rest of the array handling code was written with the implicit assumption that the arrays were non-empty.  Also, whenever someone removes defensive code in favour of handling a larger set of programs, I like to make sure it isn't a case of "the fire alarms were annoying so we turned them off".  This could quite well be the right thing to do; I'd just like to know why.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: b06a1f0415c789d7068299ccda5319162a55574c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84267309
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: jbmc/regression/jbmc-strings/cprover/CProverString.java [hunk 3]
```diff
@@ -63,6 +63,14 @@ public static StringBuilder deleteCharAt(StringBuilder sb, int index) {
         return sb.deleteCharAt(index);
     }
 
+    public static boolean equals(String a, String b)
```
Comment (peterschrammel) [inline jbmc/regression/jbmc-strings/cprover/CProverString.java]: Why is this required? Can't you use java-models-library instead?

File: src/solvers/flattening/boolbv_array.cpp (hunk 2 above)
Comment (peterschrammel) [reply src/solvers/flattening/boolbv_array.cpp:18]: > This could quite well be the right thing to do; I'd just like to know why.
> Is an empty string actually an empty array? 

In a `{ int length; char **data }` representation, what else than `{ .length = 0, .data = &_data }, _data = {}` should an empty string be? Shall data be special-cased to a null pointer to avoid zero-length arrays?

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -1849,8 +1844,8 @@ static exprt instantiate(
   for(const auto &index : indexes)
   {
     const exprt univ_var_value =
-      compute_inverse_function(stream, axiom.univ_var, val, index);
-    implies_exprt instance(
+      compute_inverse_function(axiom.univ_var, val, index);
+    exprt instance = implies_exprt(
```
Comment (majakusber) [inline src/solvers/refinement/string_refinement.cpp]: Why did this change?

File: jbmc/regression/jbmc-strings/cprover/CProverString.java (hunk 3 above)
Comment (majakusber) [inline jbmc/regression/jbmc-strings/cprover/CProverString.java]: Why is this here in the regression folder? We are loading models library so I thought the method is implemented there now.

File: src/solvers/flattening/boolbv_array.cpp (hunk 2 above)
Comment (majakusber) [reply src/solvers/flattening/boolbv_array.cpp:18]: Has this been investigated?

**Review by peterschrammel — APPROVED** [review]

File: jbmc/regression/jbmc-strings/StringBuilderInsert/test.desc [hunk 5]
```diff
@@ -1,6 +1,6 @@
 CORE
 Test.class
---function Test.check
+--function Test.check --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
 ^EXIT=10$
```
Comment (mgudemann) [inline jbmc/regression/jbmc-strings/StringBuilderInsert/test.desc:4]: does this work on windows ?

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 29bbb88a41e8a346290ada05c494b96e51a270ff).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84406021
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/solvers/refinement/string_refinement.cpp [hunk 6]
```diff
@@ -1836,12 +1836,8 @@ static exprt instantiate(
   const exprt &str,
   const exprt &val)
 {
-  const auto indexes = find_indexes(axiom.body, str, axiom.univ_var);
-  INVARIANT(
-    str.id() == ID_array || indexes.size() <= 1,
-    "non constant array should always be accessed at the same index");
   exprt::operandst conjuncts;
-  for(const auto &index : indexes)
+  for(const auto &index : find_indexes(axiom.body, str, axiom.univ_var))
```
Comment (majakusber) [inline src/solvers/refinement/string_refinement.cpp:1844]: I'm not sure if I understood this correctly but if there is more than one index via which we access `str`, all of them would get replaced with the same new index `var` which is not correct?

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 7a8e8cbc5386e2468c41fc9e6da965595fd5240a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85036174
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 1f004034a0e65210519884660d8a11b9fcc371f2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85385375
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 365d8ca40d9997cbcc485dd1c71f7a06ef57e17c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85407757
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: jbmc/regression/jbmc-strings/cprover/CProverString.java (hunk 3 above)
Comment (romainbrenguier) [reply jbmc/regression/jbmc-strings/cprover/CProverString.java]: Yes that's right, I will remove that part. :eagle:

Comment (romainbrenguier) [reply jbmc/regression/jbmc-strings/cprover/CProverString.java]: :eagle:

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 408cba983848eb5edfbbb529bee7f218bce7e6cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86245748
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: a5cce052b9502bbbf1da6615343aa38129d08f2d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86248971
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/solvers/refinement/string_refinement.cpp (hunk 6 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement.cpp:1844]: If there are several indexes then we get the conjunction of the replacement. 
For instance for a formula `forall q. s[q+x]='a' && s[q]=c` we would get `s[v] = 'a' && s[v-x] = c && s[v+x] = 'a' && s[v] = c`.
I will add this example to the doxygen.

File: src/solvers/refinement/string_refinement.cpp (hunk 4 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement.cpp]: I removed this change

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: fa07dc702a3b8dac3c67dc5b69cfc9ead76693da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86257728
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/solvers/flattening/boolbv_array.cpp (hunk 2 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_array.cpp:18]: Re @peterschrammel's comment: in C, an empty string is `{ '\0' }`, and thus the, e.g., the `string_abstractiont` code distinguishes the "string length" (which might be zero) from the size of the underlying buffer (which should be non-zero). Of course, however, we are talking about a piece of code that should _not_ be specific to C, so all this might not apply.

Comment (peterschrammel) [reply src/solvers/flattening/boolbv_array.cpp:18]: Maybe we should ask the question differently: 'What is the error case that above conditional is supposed to detect?'

Comment (tautschnig) [reply src/solvers/flattening/boolbv_array.cpp:18]: `boolbv_width` would return `0` for an array of non-constant size. That, however, seems like a bogus case here as a concrete array for a type of unknown size is (at least to me) hard to imagine. So it's probably true that the only reasonable case is that of zero operands. Would any call site of this function fail when receiving an empty array? It's hard to tell, but as of 70532f992716 we are in a much better shape to detect those.

File: jbmc/regression/jbmc-strings/StringBuilderInsert/test.desc (hunk 5 above)
Comment (romainbrenguier) [reply jbmc/regression/jbmc-strings/StringBuilderInsert/test.desc:4]: I think format_classpath.sh is made specifically to work on linux, mac and windows (unlike using `:` as a separator)

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 9f3d360cb488e6842b61f4f551584a4456e3ad83).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86437726
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by majakusber — APPROVED** [review]

**Comment by romainbrenguier** [comment]
submodule update now pointing to master

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 87bd027e8e6423bcd387af5abf2a99ecac8eda76).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86492208
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
